packages/hdwallet-native/src/native.ts

```diff
@@ -7,10 +9,57 @@ import { MixinNativeBTCWallet, MixinNativeBTCWalletInfo } from "./bitcoin";
 import { MixinNativeETHWalletInfo, MixinNativeETHWallet } from "./ethereum";
 import { MixinNativeCosmosWalletInfo, MixinNativeCosmosWallet } from "./cosmos";
 import { MixinNativeBinanceWalletInfo, MixinNativeBinanceWallet } from "./binance";
+import type { NativeAdapterArgs } from "./adapter";
+
+export enum NativeEvents {
+  MNEMONIC_REQUIRED = "MNEMONIC_REQUIRED",
+  READY = "READY",
+}
+
+export class NativeHDWalletBase {
+  readonly #events: EventEmitter2;
+
+  constructor() {
+    this.#events = new EventEmitter2();
+  }
+
+  get events() {
+    return this.#events;
+  }
+
+  /**
+   * Wrap a function call that needs a mnemonic seed
+   * @param hasMnemonic - A
+   * @param callback
+   */
+  needsMnemonic<T>(hasMnemonic: boolean, callback: () => T): T {
+    if (hasMnemonic) {
+      return callback();
+    }
+
+    this.#events.emit(
+      NativeEvents.MNEMONIC_REQUIRED,
+      core.makeEvent({
+        message_type: NativeEvents.MNEMONIC_REQUIRED,
+        from_wallet: true,
+      })
+    );
+
+    throw new Error("Wallet is not initialized");
```

Comment:
You mentioned the throw being problematic, should we just emit the event?

Reply:
there's only one place that uses the synchronous version of the function and I didn't notice any problems with that, but I didn't do a ton of testing. I'll take a look at this again.